Introduce hard-limits on the size of dispute extrinsics, and remove culprits requirement (#525)
gavofyork/graypaper: text/judgments.tex

\subsection{Extrinsic}

The disputes extrinsic $\xtdisputes$ is functional grouping of three otherwise independent extrinsics. It comprises \emph{verdicts} $\xtverdicts$, \emph{culprits} $\xtculprits$, and \emph{faults} $\xtfaults$. Verdicts are a compilation of judgments coming from exactly two-thirds plus one of either the active validator set or the previous epoch's validator set, \ie the Ed25519 keys of $\activeset$ or $\previousset$. Culprits and faults are proofs of the misbehavior of one or more validators, respectively either by guaranteeing a work-report found to be invalid, or by signing a judgment found to be in contradiction to a work-report's validity. Both of these are considered a kind of \emph{offense}. Formally:
\begin{equation}
  \begin{aligned}
    \xtdisputes &\equiv \tuple{\xtverdicts, \xtculprits, \xtfaults} \\
    \where \xtverdicts &\in \sequence[:\Cmaxextrinsicverdicts]{\tuple{
      \hash,
      \ffrac{\thetime}{\Cepochlen} - \N_2,

>       \sequence[\floor{\twothirds\valcount} + 1]{\tuple{

Reviewer: Revert this change? Not sure it is well-defined as-is (`\valcount` is a set) and it's not necessary. The 2/3+1 requirement is now specified below, as it depends on the length of the validator set that is selected in the verdict.

FWIW I also intentionally moved the "exactly two-thirds plus one" text from above to the paragraph below, so that the equations would correspond to the wording above them.

Author: Thanks. Although I tried to isolate my changes as best I could, I did not notice the change in the `exact` wording. I am going to revert both of those changes.

Offender signatures must be similarly valid and reference work-reports with judgments and may not report keys which are already in the punish-set:
\begin{align}

>   \label{eq:culprits}\forall \tup{\xc¬reporthash, \xc¬offenderindex, \xc¬signature} &\in \xtculprits : \bigwedge \abracegroup{

Reviewer: These labels not referenced anywhere now.

  \end{split}
\end{align}

> There are some logical consequences derived from the composition of this extrinsic: any verdict containing solely valid judgments implies the existence of at least one validator's vote against it that must have triggered the dispute, which in turn implies a fault; and any verdict containing solely invalid judgments implies the existence of at least two guarantees signed for that report, which in turn implies the existence of at least two culprits. 

Reviewer: ```suggestion
Any verdict containing solely valid judgments implies the existence of at least one validator's vote against it that must have triggered the dispute, which in turn implies a fault; and any verdict containing solely invalid judgments implies the existence of at least two guarantees signed for that report, which in turn implies the existence of at least two culprits. 
```

The "derived from the composition of this extrinsic" bit doesn't make sense to me; these are more logical consequences of how auditing works?

> Faults and culprits must never list keys that have already been included in the punish-set following some previously seen offence (equations \ref{eq:culprits} and \ref{eq:faults}), and so although logically these minimum offenses must exist for each corresponding verdict, they may not be actually included in a faults or culprits extrinsic: good verdicts should only be generated when there is at least one associated new fault offender; bad verdicts should always be generated even when they don't have any new associated culprit offender since we need to exclude bad reports.

Reviewer: ```suggestion
Faults and culprits must never list keys that have already been included in the punish-set following some previously reported offence (equations \ref{eq:culprits} and \ref{eq:faults}), and so although logically these minimum offenses must exist for each corresponding verdict, it may not be possible to include them in the extrinsic. Good verdicts should only be generated when there is at least one associated new fault offender, as they otherwise would have no effect. Bad verdicts should always be generated even when they don't have any new associated culprit offender since we need to exclude bad reports.
```

Formally:
\begin{align}

>   \forall \tup{\¬reporthash, \floor{\twothirds\valcount} + 1} \in \verdicts &:

Reviewer: `\floor{\twothirds\valcount} + 1` should be `\top` now.

Author: Thanks. Fixing.

    \exists \tup{\¬reporthash, \dots} \in \xtfaults
\end{align}

> Although this should not be expected in practice, in theory it is possible that the number of offenses could be large (up to $\onethird \valcount$ faults and up to three culprits for each bad verdict). In order not to burden block creation overmuch, we specify practical limits to the number of verdicts and offenses which can be included in a single block.

Reviewer: Again `\onethird \valcount` doesn't really make sense, `\valcount` is the set of possible validator counts.

Author: Ok, rewording to use "$\onethird$ of validators"

> Formally,

Reviewer: These equations are redundant as the limits are specified in the first equation in this section. Probably all this explanatory text about limits is not really necessary either, though it's also probably harmlesss and it seems accurate to me.

Author: I'll remove the equations then. I'm keen to leave the text in, though, but make it point to equation 10.2

  \end{split}
\end{align}

> There are some constraints placed on the composition of this extrinsic: any verdict containing solely valid judgments implies the same report having at least one valid entry in the faults sequence $\xtfaults$. Good verdicts should only be reported when there is at least one associated new fault offender, as they otherwise would have no effect. Bad verdicts should always be reported.

Reviewer: ```suggestion
Any verdict containing solely valid judgments implies the same report having at least one valid entry in the faults sequence $\xtfaults$. Formally:
```

The second sentence seems to just be a rewording of the first. Also wondering what the reason is for moving this from where it was before?

Formally:
\begin{align}
  \forall \tup{\¬reporthash, \top} \in \verdicts &:
    \exists \tup{\¬reporthash, \dots} \in \xtfaults
\end{align}

> Although this should not be expected in practice, in theory it is possible that the number of offenses could be large (up to $\onethird$ of validators as faults and up to three culprits for each bad verdict). In order not to over burden block creation, we specify practical limits to the number of verdicts and offenses which can be included in a single block as can be seen in equation \ref{eq:disputesextrinsics}.

Reviewer: Would probably be clearer to put this text by the equations that formalise it. Personally I would just remove it as I don't think it adds much.

  \end{split}
\end{align}

> Any verdict containing solely valid judgments implies the existence of at least one validator's vote against it that must have triggered the dispute, which in turn implies a fault; and any verdict containing solely invalid judgments implies the existence of at least two guarantees signed for that report, which in turn implies the existence of at least two culprits. 

Reviewer 2: This seems like a less clear rewrite of my previous paragraph: talking about "validators vote" as a separate consideration, but strictly implying, a "fault" is nonsense. Stick with the original wording.

> Faults and culprits must never list keys that have already been included in the punish-set following some previously reported offence (equations \ref{eq:culprits} and \ref{eq:faults}), and so although logically these minimum offenses must exist for each corresponding verdict, it may not be possible to include them in the extrinsic. Good verdicts should only be generated when there is at least one associated new fault offender, as they otherwise would have no effect. Bad verdicts should always be generated even when they don't have any new associated culprit offender since we need to exclude bad reports.

Reviewer 2: Is this meant to be another paragraph? If so it needs a newline before the last one.

Formally:
\begin{align}
  \forall \tup{\¬reporthash, \top} \in \verdicts &:
    \exists \tup{\¬reporthash, \dots} \in \xtfaults
\end{align}

> Although this should not be expected in practice, in theory it is possible that the number of offenses could be large (up to $\onethird$ of validators as faults and up to three culprits for each bad verdict). In order not to burden block creation overmuch, we specify practical limits to the number of verdicts and offenses which can be included in a single block as can be seen in equation \ref{eq:disputesextrinsics}.

Reviewer 2: ```suggestion
Although this should not be expected in practice, in theory it is possible that the number of offenses could be large (up to $\onethird$ of validators as faults and up to three culprits for each bad verdict). In order not to over burden block creation, we specify practical limits to the number of verdicts and offenses which can be included in a single block as can be seen in equation \ref{eq:disputesextrinsics}.
```
